src/ott/geometry/geodesic.py

```diff
@@ -0,0 +1,260 @@
+      **kwargs: Any
+  ):
+    super().__init__(epsilon=1., **kwargs)
+    self.laplacian = laplacian
+    self.t = t
+    self.order = order
+    self.chebyshev_coeffs = chebyshev_coeffs
+    self.numerical_scheme = numerical_scheme
+
+  @classmethod
+  def from_graph(
+      cls,
+      G: jnp.ndarray,
+      t: Optional[float] = 1e-3,
+      order=100,
+      directed: bool = False,
+      normalize: bool = False,
+      **kwargs: Any
+  ) -> "Geodesic":
+    r"""Construct a Geodesic geometry from an adjacency matrix.
+
+    Args:
+      G: Adjacency matrix.
+      t: Time parameter for approximating the geodesic exponential kernel.
+        If `None`, it defaults to :math:`\frac{1}{|E|} \sum_{(u, v) \in E}
+        \text{weight}(u, v)` :cite:`crane:13`. In this case, the ``graph``
+        must be specified and the edge weights are assumed to be positive.
+      order: Max order of Chebyshev polynomial.
+      directed: Whether the ``graph`` is directed. If not, it's made
+        undirected as :math:`G + G^T`. This parameter is ignored when passing
+        the Laplacian directly, assumed to be symmetric.
+      normalize: Whether to normalize the Laplacian as
+        :math:`L^{sym} = \left(D^+\right)^{\frac{1}{2}} L
+        \left(D^+\right)^{\frac{1}{2}}`, where :math:`L` is the
+        non-normalized Laplacian and :math:`D` is the degree matrix.
+      kwargs: Keyword arguments for the Geodesic class.
+
+    Returns:
+      The Geodesic geometry.
+    """
+    assert G.shape[0] == G.shape[1], G.shape
+
+    if directed:
+      G = G + G.T
+
+    degree = jnp.sum(G, axis=1)
+    laplacian = jnp.diag(degree) - G
+
+    if normalize:
+      inv_sqrt_deg = jnp.diag(
+          jnp.where(degree > 0.0, 1.0 / jnp.sqrt(degree), 0.0)
+      )
+      laplacian = inv_sqrt_deg @ laplacian @ inv_sqrt_deg
+
+    if t is None:
+      t = (jnp.sum(G) / jnp.sum(G > 0.)) ** 2
+
+    # Create an instance of the Geodesic class and set the attribute
+    geodesic_instance = cls(laplacian, t=t, order=order, **kwargs)
+
+    # Compute the coeffs of the Chebyshev pols approx using Bessel functs.
+    chebyshev_coeffs = (
+        2 *
+        ive(jnp.arange(0, geodesic_instance.order + 1), -geodesic_instance.t)
+    ).tolist()
+
+    # Set the attribute
+    geodesic_instance.chebyshev_coeffs = chebyshev_coeffs
+
+    return geodesic_instance
+
+  def apply_kernel(
+      self,
+      scaling: jnp.ndarray,
+      eps: Optional[float] = None,
+      axis: int = 0,
+  ) -> jnp.ndarray:
+    r"""Apply :attr:`kernel_matrix` on positive scaling vector.
+
+    Args:
+    scaling: Scaling to apply the kernel to.
+    eps: passed for consistency, not used yet.
+    axis: passed for consistency, not used yet.
+
+    Returns:
+    Kernel applied to ``scaling``.
+    """
+
+    def compute_largest_eigenvalue(laplacian_matrix, k, seed=None):
+      # Compute the largest eigenvalue of the Laplacian matrix.
+      if seed is None:
+        seed = jax.random.PRNGKey(0)
```

Comment:
When refactoring this, use `from ott import utils; utils.default_prng_key`.

Reply:
Fixed.